modules/3d/src/rendering.cpp

```diff
@@ -0,0 +1,367 @@
+        {
+            if (validPtr[x])
+            {
+                float d = inp[x];
+                // precision-optimized version of this:
+                //float z = - zFar * zNear / (d * (zFar - zNear) - zFar);
+                float z =  1.f / ((1.f - d) * scaleNear + d * scaleFar );
+                outp[x] = z;
+            }
+        }
+    }
+}
+
+// [zNear, zFar] -> [0, 1]
+static void invertDepth(const Mat& inbuf, Mat& outbuf, Mat& validMask, double zNear, double zFar)
+{
+    CV_Assert(inbuf.type() == CV_32FC1);
+    outbuf.create(inbuf.size(), CV_32FC1);
+    validMask.create(inbuf.size(), CV_8UC1);
+
+    float fNear = (float)zNear, fFar = (float)zFar;
+    float zadd = (float)(zFar / (zFar - zNear));
+    float zmul = (float)(-zNear * zFar / (zFar - zNear));
+    for (int y = 0; y < inbuf.rows; y++)
+    {
+        const float * inp = inbuf.ptr<float>(y);
+        float * outp = outbuf.ptr<float>(y);
+        uchar * validPtr = validMask.ptr<uchar>(y);
+        for (int x = 0; x < inbuf.cols; x++)
+        {
+            float z = inp[x];
+            uchar m = (z >= fNear) && (z <= fFar);
+            z = std::max(std::min(z, fFar), fNear);
+            // precision-optimized version of this:
+            // outp[x] = (z - zNear) / z * zFar / (zFar - zNear);
+            outp[x] = zadd + zmul / z;
+            validPtr[x] = m;
+        }
+    }
+}
+
+
+
+CV_EXPORTS  void triangleRasterize(InputArray _vertices, InputArray _indices, InputArray _colors,
+                                   InputArray cameraPose, double fovyRadians, double zNear, double zFar,
+                                   TriangleRasterizeSettings settings,
+                                   InputOutputArray _depthBuffer, InputOutputArray _colorBuffer)
+{
+    CV_Assert(cameraPose.type() == CV_32FC1 || cameraPose.type() == CV_64FC1);
+    CV_Assert((cameraPose.size() == Size {4, 3}) || (cameraPose.size() == Size {4, 4}));
+
+    CV_Assert((fovyRadians > 0) && (fovyRadians < CV_PI));
+    CV_Assert(zNear > 0);
+    CV_Assert(zFar > zNear);
+
+    Mat cpMat;
+    cameraPose.getMat().convertTo(cpMat, CV_64FC1);
+    Matx44d camPoseMat = Matx44d::eye();
+    for (int i = 0; i < 3; i++)
+    {
+        for (int j = 0; j < 4; j++)
+        {
+            camPoseMat(i, j) = cpMat.at<double>(i, j);
+        }
+    }
+
+    bool hasIdx    = !_indices.empty();
+    bool hasColors = !_colors.empty();
+
+    Mat vertices, colors, triangles;
+    int nVerts = 0, nColors = 0, nTriangles = 0;
+
+    if (hasIdx)
+    {
+        CV_CheckFalse(_vertices.empty(), "No vertices provided along with indices array");
+
+        int vertexType = _vertices.type();
+        CV_Assert(vertexType == CV_32FC1 || vertexType == CV_32FC3);
+        vertices = _vertices.getMat();
+        if ((_vertices.channels() == 1) && (_vertices.rows() == 3))
+        {
+            vertices = vertices.t();
+        }
+        vertices = vertices.reshape(3, 1).t();
```

Comment:
.t() may happen twice.

Reply:
Already [discussed](https://github.com/opencv/opencv/pull/24459#discussion_r1456037998):
> That second transposition is performed on 1xN matrix so it's almost free in terms of performance
Since it's not obvious, I'll add an explaining comment.